include/Zydis/Encoder.h

```diff
@@ -0,0 +1,440 @@
+                                                       (ZYDIS_ENCODABLE_ENCODING_MVEX - 1)),
+    /**
+     * The minimum number of bits required to represent all values of this enum.
+     */
+    ZYDIS_ENCODABLE_ENCODING_REQUIRED_BITS          = 
+        ZYAN_BITS_TO_REPRESENT(ZYDIS_ENCODABLE_ENCODING_MAX_VALUE)
+} ZydisEncodableEncoding;
+
+/**
+ * Defines user-encodable instruction prefixes. Some prefixes e.g. address/operand size overrides
+ * cannot be requested explicitly. It's encoder's responsibility to use them as needed.
+ */
+typedef enum ZydisEncodablePrefix_
+{
+    ZYDIS_ENCODABLE_PREFIX_NONE                     = 0x00000000,
+    ZYDIS_ENCODABLE_PREFIX_LOCK                     = 0x00000001,
+    ZYDIS_ENCODABLE_PREFIX_REP                      = 0x00000002,
+    ZYDIS_ENCODABLE_PREFIX_REPE                     = 0x00000004,
+    ZYDIS_ENCODABLE_PREFIX_REPZ                     = ZYDIS_ENCODABLE_PREFIX_REPE,
+    ZYDIS_ENCODABLE_PREFIX_REPNE                    = 0x00000008,
+    ZYDIS_ENCODABLE_PREFIX_REPNZ                    = ZYDIS_ENCODABLE_PREFIX_REPNE,
+    ZYDIS_ENCODABLE_PREFIX_BND                      = 0x00000010,
+    ZYDIS_ENCODABLE_PREFIX_XACQUIRE                 = 0x00000020,
+    ZYDIS_ENCODABLE_PREFIX_XRELEASE                 = 0x00000040,
+    ZYDIS_ENCODABLE_PREFIX_BRANCH_NOT_TAKEN         = 0x00000080,
+    ZYDIS_ENCODABLE_PREFIX_BRANCH_TAKEN             = 0x00000100,
+    ZYDIS_ENCODABLE_PREFIX_NOTRACK                  = 0x00000200,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_CS               = 0x00000400,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_SS               = 0x00000800,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_DS               = 0x00001000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_ES               = 0x00002000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_FS               = 0x00004000,
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS               = 0x00008000,
+
+    ZYDIS_ENCODABLE_PREFIX_SEGMENT_MASK             = (ZYDIS_ENCODABLE_PREFIX_SEGMENT_CS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_SS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_DS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_ES |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_FS |
+                                                       ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS),
+
+    /**
+     * Maximum value of this enum.
+     */
+    ZYDIS_ENCODABLE_PREFIX_MAX_VALUE                = (ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS |
+                                                       (ZYDIS_ENCODABLE_PREFIX_SEGMENT_GS - 1)),
+    /**
+     * The minimum number of bits required to represent all values of this enum.
+     */
+    ZYDIS_ENCODABLE_PREFIX_REQUIRED_BITS            =
+        ZYAN_BITS_TO_REPRESENT(ZYDIS_ENCODABLE_PREFIX_MAX_VALUE)
+} ZydisEncodablePrefix;
```

Comment:
Would it make sense to also use `ZYDIS_ATTRIB_` here and just define a mask of the attributes that users can explicitly request? Arguably, it might make it a harder for users to determine which attributes they can explicitly request. On the positive side, it'd simplify `ZydisEncoderDecodedInstructionToEncoderRequest` and reduce duplication. Not really sure what's the best way here.

Reply:
Isn't that a duplicate of previous comment? I will work on this item as soon as I finish size-optimal outputs and regression test suite. I'm not sure about other API change though (`branch_type`).

Reply:
The previous comment was on `ZydisEncodableAttribute`, this one is on `ZydisEncodablePrefix`. However, if it's any consolation, I also completely mixed them up. In fact, I originally wanted to comment it on `ZydisEncodablePrefix`, accidentally commenting on `ZydisEncodableAttribute` instead! 😅 Fortunately for me, it also made sense in the other context.

Reply:
Oh, in my head they are a single type anyway :D `ZYDIS_ATTRIB_` will replace both :)